Encapsulate KCM controllers with their metadata

atiratree opened the pull request:
#### What type of PR is this?

/kind cleanup

#### What this PR does / why we need it:

this is a followup to following PR and a discussion: https://github.com/kubernetes/kubernetes/pull/115813#discussion_r1214768774


- These metadata can be used to handle controllers in a generic way.
- This enables showing feature gated controllers in kube-controller-manager's help.
- It is possible to obtain a controllerName in the InitFunc so it can be passed down to and used by the controller.
- include ServiceAccountTokenController in the NewRegistrableControllers to make it more generic
- move start controller pre- and post- checks/actions out of StartControllers into StartController function


```
metadata about a controller:
- name
- requiredFeatureGates
- aliases
- isDisabledByDefault
- isCloudProviderController
- requiresSpecialHandling
```


#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #

#### Special notes for your reviewer:

similar thing can be done for cloud-controller-manager, but it is better to open a new separate PR for that

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
kube-controller-manager's help will include controllers behind a feature gate in `--controllers` flag
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

atiratree committed:
Encapsulate KCM controllers with their metadata

- These metadata can be used to handle controllers in a generic way.
- This enables showing feature gated controllers in kube-controller-manager's help.
- It is possible to obtain a controllerName in the InitFunc so it can be passed down to and used by the controller.

metadata about a controller:
- name
- requiredFeatureGates
- isDisabledByDefault
- isCloudProviderController

atiratree committed:
include ServiceAccountTokenController in the NewControllerDescriptors to make it more generic

- pass a map of controllerDescriptors instead of a function

atiratree committed:
move aliases into each registrable controller

atiratree commented on cmd/kube-controller-manager/app/controllermanager.go:
```diff
@@ -389,100 +397,183 @@ func (c ControllerContext) IsControllerEnabled(name string) bool {
-type InitFunc func(ctx context.Context, controllerCtx ControllerContext) (controller controller.Interface, enabled bool, err error)
+type InitFunc func(ctx context.Context, controllerContext ControllerContext, controllerName string) (controller controller.Interface, enabled bool, err error)
+
+type RegistrableController interface {
```
- can anybody come up with a better name?
- should I move this to a separate file?

thockin replied:
Yeah, I don't love the name.

`ControllerDescriptor` or something?  I am bad at naming things.

thockin replied:
Why do we need this as an interface?  For testing or will there be real implementations?

thockin replied:
I think simpler is best, so a struct with public members (no interface) plus well understood functions.  You could put this in a new file or even in sub-package.

atiratree replied:
My main concern was to allow people to inject a custom logic into the function calls if needed. But I suppose, since we do not have no such use case yet, we could do without an interface.

It is hard to come up with a good name for this one.  Not sure about `Descriptor` yet, because it also contains the logic for running the controllers... 

Also not sure how to name the functions: for now I have `newEndpointsRegistrableController` for the following reasons:
1. should not refer to creating or starting the controller, since that is done elsewhere
2. should not have a long prefix so the original controller name is easily readable, so suffix is preferable IMO

thockin replied:
This is not a public API (at least not outside of k/k) so we can always change it in the event we really do need an interface.  But it seems like a "dumb data object" is appropriate.  Even with a function pointer.

I like `Descriptor` or `Handle` because this object is the beginning of a controller.  It's not running yet.  It holds metadata, but we have too many `*Meta` things.

Alternately, you could make this all part of `controller.Interface` and document that lifecycle as having a pre-running phase.  That spreads out the metadata more and puts it near the controller implementation, which I am not sure is right.  This feels like an "inode vs. file" thing - many inodes (names and aliases) can refer to the same file (descriptor).  That said, the only method in `controller.Interface` is `Name()` and it seems like many of these `start` functions return `nil` (e.g. `startServiceController`) for the controller.Interface anyway!!  That should be a cleanup of its own.

`newEndpointsControllerDescriptor` is long but not terrible.  It says what it means.  `newEndpointsControllerDesc` is a bit shorter and not much less readable.  Either one satisfies your criteria.  

So I would say either something like

```
func newFooControllerDescriptor() ControllerDescriptor {
    return ControllerDescriptor{
        Name: "foo",
        Aliases: []string{"foo-ctrl"},
        Run: startFooController,
    }
}

func startFooController() error {
    // lots of stuff
}
```

or else bake this into `controller.Interface` like:

```
type Interface interface {
    Name() string
    Aliases() []string
    Run(ctrlCtxControllerContext ) error
}
````

and then just call `NewFooController() controller.Interface`, check gates, then call the Run() method on it.

atiratree replied:
I suppose we can look at it as a descriptor - I like the suggestions. I would prefer we leave `controller.Interface` refactoring out of this PR, and I am not yet convinced that we should move the functionality there. Nevertheless, there should not be too many changes in the code that this PR touches, if we decide to move it.

atiratree replied:
renamed to ControllerDescriptor

atiratree commented on cmd/kube-controller-manager/app/controllermanager.go:
```diff
@@ -572,16 +670,39 @@ func StartControllers(ctx context.Context, controllerCtx ControllerContext, star
 	//   and those emitted for it - this is a bit debatable and could be revised.
-	for controllerName, initFn := range controllers {
-		if !controllerCtx.IsControllerEnabled(controllerName) {
+	for controllerName, regController := range registrableControllers {
```
should I make a function from a body of this for so it can be reused above for `ServiceAccountTokenController`?

thockin replied:
Smaller functions are better.

atiratree replied:
the size most likely won't be that much changed, but I think the biggest benefit would be the deduplication and price small indirection

atiratree replied:
deduplicated

atiratree commented on cmd/kube-controller-manager/app/controllermanager.go:
```diff
@@ -619,20 +740,22 @@ func StartControllers(ctx context.Context, controllerCtx ControllerContext, star
-type serviceAccountTokenControllerStarter struct {
-	rootClientBuilder clientbuilder.ControllerClientBuilder
+// RegistrableController map so that it can always run first.
+func newServiceAccountTokenRegistrableController(rootClientBuilder clientbuilder.ControllerClientBuilder) *registrableController {
```
can we move this controller to a new file?

atiratree commented:
cc people from the last PR: @liggitt @nckturner @aojea @cici37 @alexanderConstantinescu  @msau42 @jprzychodzen @soltysh @andrewsykim @cheftako  @thockin @sftim

thockin reviewed (commented):
This was just a skim, but I personally like the direction.  I'd emphasize the encapsulation by segregating this code behind a clean API (even if in the same pkg, but a different file) and making the code that calls it as simple and streamlined as possible.

alexzielenski commented:
/triage accepted

atiratree committed:
move start controller pre- and post- checks/actions out of StartControllers

into StartController function

the function is reused by ServiceAccountTokenController

atiratree commented:
as a result of the discussions, I have made the following changes: 

- renamed `RegistrableController` to `ControllerDescriptor`
- removed `ControllerDescriptor` interface in favor of a struct
- moved start controller pre- and post- checks/actions out of `StartControllers` into `StartController` function

aojea commented on cmd/kube-controller-manager/app/validatingadmissionpolicystatus.go:
```diff
@@ -21,13 +21,26 @@ import (
+	return &ControllerDescriptor{
+		name:     names.ValidatingAdmissionPolicyStatusController,
+		initFunc: startValidatingAdmissionPolicyStatusController,
+		requiredFeatureGates: []featuregate.Feature{
```
I don´t have the feeling this is a good idea, because right now it looks pretty neat and clean, but as soon as the code evolves and new features are added this is just yet another place to update the feature gates ... or what happen if the controller has only one part of the logic with a feature gate? we need to run the controller and we still need to add a feature gate inside the controller logic

some food for thoughts

atiratree replied:
It depends on what tradeoff we want to make. If we want to have all the controller metadata and logic sprinkled everywhere or together. We already have feature gate logic sprinkled everywhere, so IMO this would not change that much. That is, one has to search for all occurrences of a feature gate to get the full picture.

 > or what happen if the controller has only one part of the logic with a feature gate? we need to run the controller and we still need to add a feature gate inside the controller logic

I think we would solve this the same way we do today. We can put a feature gate call inside the init function or inside the controller itself.

For the usual cases, we could reason about controllers without running them. That is, we could for example show which controllers are gated by which feature gates in kube-controller-manager's help

aojea replied:
the logic always going to be sprinkled everywhere, the feature gated code can be anything, a function or just part of a function ... 

> For the usual cases, we could reason about controllers without running them. That is, we could for example show which controllers are gated by which feature gates in kube-controller-manager's help

I think that the intent is good, but I'm afraid that the reality is different and it soon will get out of sync and we'll have an additional place interacting with the code that is not obvious for everybody

atiratree replied:
I know it only helps a little but I have added a unit test to make sure people do not mix the old and the new approach.

atiratree committed:
add unit tests for NewControllerDescriptors

- controller descriptors should not be feature gated
- aliases should not be defined for new controllers and have only a
  canonical name

atiratree commented:
https://github.com/kubernetes/kubernetes/pull/119208 shows a use case for this refactoring and https://github.com/kubernetes/kubernetes/pull/119208#discussion_r1352390164 discusses how controller name should be consumed.

atiratree commented:
rebased

thockin commented on cmd/kube-controller-manager/app/controllermanager.go:
```diff
@@ -388,101 +390,141 @@ func (c ControllerContext) IsControllerEnabled(name string) bool {
+// ControllerDescriptorsFunc is used to create a collection of controller descriptors
 // given the loopMode.
-type ControllerInitializersFunc func(loopMode ControllerLoopMode) (initializers map[string]InitFunc)
+type ControllerDescriptorsFunc func() (initializers map[string]*ControllerDescriptor)
```
"initializers" seems like a weird name for this - why name the return at all?

atiratree replied:
True. Although I have noticed that after the changes in the https://github.com/kubernetes/kubernetes/pull/120371/commits/a85779b4dfe4c71a2def2f2385fb4d8f896522ad commit, we do not need the function type at all. So I have removed it.

thockin commented on cmd/kube-controller-manager/app/controllermanager.go:
```diff
@@ -619,6 +631,10 @@ func StartControllers(ctx context.Context, controllerCtx ControllerContext, star
 	// - it allows distinguishing between log entries emitted by the controller
 	//   and those emitted for it - this is a bit debatable and could be revised.
 	for controllerName, controllerDesc := range controllerDescriptors {
+		if controllerDesc.RequiresSpecialHandling() {
```
I expected to see a similar loop filtering for `RequiresSpecialHandling()` and running those, but there's only one such and it's hard-coded above - why not 

```
for _,  doingSpecialHandling := range []bool{true, false} {
    for controllerName, controllerDesc := range controllerDescriptors {
        if controllerDesc.RequiresSpecialHandling() != doingSpecialHandling {
            continue
        }
    //...
}
```

atiratree replied:
@thockin 

1. I am worried that with this approach, there might be a new controller in the future that needs special handling and would start before ServiceAccountTokenController which we expect to be run first. I thought about adding ordering for the controllers or a graph of dependencies, but that seems like an overkill for now IMO, since we have just one use case.
2. we also have the following line where we initialize cloud provider after the ServiceAccountTokenController is started.
```
	controllerCtx.Cloud.Initialize(controllerCtx.ClientBuilder, ctx.Done())
```

thockin reviewed (commented):
really minor questions

thockin commented:
Thanks!

/lgtm
/approve

atiratree commented:
Thanks for the review @thockin. Unfortunately I missed controllers enabled by default in one test case. And just ran into it during a review https://github.com/kubernetes/kubernetes/pull/119208#discussion_r1376171223. 

I have opened a small PR for this: https://github.com/kubernetes/kubernetes/pull/121611
